nucypher/blockchain/economics.py

```diff
@@ -14,60 +14,177 @@
+                 initial_inflation: int = __default_initial_inflation,
+                 halving_delay: int = __default_token_halving,
+                 reward_saturation: int = __default_reward_saturation,
+                 small_stake_multiplier: Decimal = __default_small_stake_multiplier
+                 ):
         """
-        Calculate parameters to use in token issuer from high-level
-        human-understandable parameters
-
-        :param initial_supply: coins at t=0,
-        :param initial_inflation: inflation in day 1 expressed in units of year**-1
-            e.g. 1/365 of all coins in day 1 is 1 year**-1
-        :param T_half: time for inflation halving in years
-        :param T_sat: "saturation" time - if staking is longer than T_sat, the
-            reward doesn't get higher
-        :param small_staker_multiplier: what fraction of max reward rate do we
-            pay to those who are about to unlock coins
+        :param initial_supply: Tokens at t=0
+        :param initial_inflation: Inflation on day 1 expressed in units of year**-1
+        :param halving_delay: Time for inflation halving in years
+        :param reward_saturation: "saturation" time - if staking is longer than T_sat, the reward doesn't get any higher
+        :param small_stake_multiplier: Fraction of maximum reward rate paid to those who are about to unlock tokens
         """
 
-        # Saving input in case we'll need it
-        self.input_params = dict(
-                initial_supply=initial_supply,
-                initial_inflation=initial_inflation,
-                T_half=T_half,
-                T_sat=T_sat,
-                small_staker_multiplier=small_staker_multiplier)
+        #
+        # Calculate
+        #
+
+        with localcontext() as ctx:
+            ctx.prec = self._precision
+
+            initial_supply = Decimal(initial_supply)
+
+            # ERC20 Token parameter (See Equation 4 in Mining paper)
+            total_supply = initial_supply * (1 + initial_inflation * halving_delay / LOG2)
+
+            # Remaining / Reward Supply - Escrow Parameter
+            reward_supply = total_supply - initial_supply
+
+            # k2 - Escrow parameter
+            staking_coefficient = 365 ** 2 * reward_saturation * halving_delay / LOG2 / (1 - small_stake_multiplier)
+
+            # k1 - Escrow parameter
+            locked_periods_coefficient = 365 * reward_saturation * small_stake_multiplier / (1 - small_stake_multiplier)
+
+            # Awarded periods- Escrow parameter
+            maximum_locked_periods = reward_saturation * 365
+
+        # Injected
+        self.initial_supply = initial_supply
+        self.initial_inflation = initial_inflation
+        self.token_halving = halving_delay
+        self.token_saturation = reward_saturation
+        self.small_stake_multiplier = small_stake_multiplier
+
+        # Calculated
+        self.__total_supply = total_supply
```

Comment:
Why keep `self.__total_supply` as private? The remaining attributes are public, and I think that's fine.

Reply:
I'm trying to discourage the accidental usage of `self.total_supply` since it needs to be case with the property below.  I can go either way on this one.